File: backend/F1Fantasy/Services/IpBlacklistService.cs
using System.Collections.Concurrent;

namespace F1Fantasy.Services;

public interface IIpBlacklistService
{
    bool IsBlacklisted(string ipAddress);
    void Blacklist(string ipAddress, string reason, TimeSpan? duration = null);
    void Unblacklist(string ipAddress);
    Dictionary<string, BlacklistEntry> GetBlacklistedIps();
}

public class BlacklistEntry
{
    public string IpAddress { get; set; } = string.Empty;
    public string Reason { get; set; } = string.Empty;
    public DateTime BlacklistedAt { get; set; }
    public DateTime? ExpiresAt { get; set; }
}

public class IpBlacklistService : IIpBlacklistService
{
    private readonly ConcurrentDictionary<string, BlacklistEntry> _blacklist = new();
    private readonly ILogger<IpBlacklistService> _logger;

    public IpBlacklistService(ILogger<IpBlacklistService> logger)
    {
        _logger = logger;
    }

    public bool IsBlacklisted(string ipAddress)
    {
        if (_blacklist.TryGetValue(ipAddress, out var entry))
        {
            // Check if temporary ban has expired
            if (entry.ExpiresAt.HasValue && entry.ExpiresAt.Value < DateTime.UtcNow)
            {
                _blacklist.TryRemove(ipAddress, out _);
                _logger.LogInformation("Temporary blacklist expired for IP: {IpAddress}", ipAddress);
                return false;
            }
            
            return true;
        }
        
        return false;
    }

    public void Blacklist(string ipAddress, string reason, TimeSpan? duration = null)
    {
        var entry = new BlacklistEntry
        {
            IpAddress = ipAddress,
            Reason = reason,
            BlacklistedAt = DateTime.UtcNow,
            ExpiresAt = duration.HasValue ? DateTime.UtcNow.Add(duration.Value) : null
        };

        _blacklist[ipAddress] = entry;
        
        var durationType = duration.HasValue ? $"for {duration.Value.TotalMinutes} minutes" : "permanently";
        _logger.LogWarning("IP {IpAddress} blacklisted {Duration}. Reason: {Reason}", 
            ipAddress, durationType, reason);
    }

    public void Unblacklist(string ipAddress)
    {
        if (_blacklist.TryRemove(ipAddress, out _))
        {
            _logger.LogInformation("IP {IpAddress} removed from blacklist", ipAddress);
        }
    }

    public Dictionary<string, BlacklistEntry> GetBlacklistedIps()
    {
        // Clean up expired entries first
        var expiredIps = _blacklist
            .Where(kvp => kvp.Value.ExpiresAt.HasValue && kvp.Value.ExpiresAt.Value < DateTime.UtcNow)
            .Select(kvp => kvp.Key)
            .ToList();

        foreach (var ip in expiredIps)
        {
            _blacklist.TryRemove(ip, out _);
        }

        return new Dictionary<string, BlacklistEntry>(_blacklist);
    }
}

File: backendJava/f1fantasy/src/main/java/no/f1fantasy/service/IpBlacklistService.java
package no.f1fantasy.service;

import org.springframework.stereotype.Service;

import java.time.OffsetDateTime;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Service
public class IpBlacklistService {

    public static class BlacklistEntry {
        private final String ipAddress;
        private final String reason;
        private final OffsetDateTime blacklistedAt;
        private final OffsetDateTime expiresAt;

        public BlacklistEntry(String ipAddress, String reason, OffsetDateTime blacklistedAt, OffsetDateTime expiresAt) {
            this.ipAddress = ipAddress;
            this.reason = reason;
            this.blacklistedAt = blacklistedAt;
            this.expiresAt = expiresAt;
        }

        public String getIpAddress() {
            return ipAddress;
        }

        public String getReason() {
            return reason;
        }

        public OffsetDateTime getBlacklistedAt() {
            return blacklistedAt;
        }

        public OffsetDateTime getExpiresAt() {
            return expiresAt;
        }
    }

    private final ConcurrentHashMap<String, BlacklistEntry> blacklist = new ConcurrentHashMap<>();

    public boolean isBlacklisted(String ipAddress) {
        BlacklistEntry entry = blacklist.get(ipAddress);
        if (entry == null) {
            return false;
        }

        if (entry.getExpiresAt() != null && entry.getExpiresAt().isBefore(OffsetDateTime.now())) {
            blacklist.remove(ipAddress);
            return false;
        }

        return true;
    }

    public void blacklist(String ipAddress, String reason, java.time.Duration duration) {
        OffsetDateTime now = OffsetDateTime.now();
        OffsetDateTime expiresAt = duration == null ? null : now.plus(duration);
        blacklist.put(ipAddress, new BlacklistEntry(ipAddress, reason, now, expiresAt));
    }

    public void blacklist(String ipAddress, String reason) {
        blacklist(ipAddress, reason, null);
    }

    public void unblacklist(String ipAddress) {
        blacklist.remove(ipAddress);
    }

    public Map<String, BlacklistEntry> getBlacklistedIps() {
        blacklist.entrySet().removeIf(entry -> entry.getValue().getExpiresAt() != null && entry.getValue().getExpiresAt().isBefore(OffsetDateTime.now()));
        return Map.copyOf(blacklist);
    }
}
